opm/simulators/wells/MultisegmentWell_impl.hpp

```diff
@@ -1303,24 +1304,9 @@ namespace Opm
     typename MultisegmentWell<TypeTag>::EvalWell
     MultisegmentWell<TypeTag>::
     getSegmentRateUpwinding(const int seg,
-                            const int comp_idx,
-                            const bool upwinding,
-                            int& seg_upwind) const
+                            const int comp_idx) const
     {
-        // not considering upwinding for the injectors for now
-        if ((!upwinding) || (well_type_ == INJECTOR) || (primary_variables_evaluation_[seg][GTotal] <= 0.) || (seg == 0)) {
-            seg_upwind = seg; // using the composition from the seg
-            return primary_variables_evaluation_[seg][GTotal] * volumeFractionScaled(seg, comp_idx);
-        }
-
-        // assert( seg != 0); // if top segment flowing towards the wrong direction, we are not handling it
-
-        // basically here, it a producer and flow is in the injecting direction
-        // we will use the compsotion from the outlet segment
-        const int outlet_segment_index = segmentNumberToIndex(segmentSet()[seg].outletSegment());
-        seg_upwind = outlet_segment_index;
-
-        // TODO: we can refactor above code to use the following return
+        const int seg_upwind = upwinding_segments_[seg];
         return primary_variables_evaluation_[seg][GTotal] * volumeFractionScaled(seg_upwind, comp_idx);
     }
```

Comment:
Maybe I am wrong, but is this correct? You are multiplying derivatives of the primary variables wrt to the segment "seg" with ones wrt. to the segment "seg_upwind"  I have not looked very carefully at the code so I may have missed something.

Reply:
Yes. We use the composition from the upwinding direction. We have some cases in the wellbore of MSW, the flow can go different directions (some flux go up, some flux go down).  Without upwinding, the flux will be with wrong composition, then the mass of component will not be conserved.